src/helper_utils.cc

```diff
@@ -0,0 +1,181 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#include "helper_utils.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <cassert>
+#include <climits>
+#include <cstdio>
+#include <cstring>
+#include <vector>
+
+#include "x509_helper_log.h"
+
+using namespace std;  // NOLINT
+
+/**
+ * For a given pid, extracts the env_name path from the foreign
+ * process' environment.  Stores the resulting path in the user provided buffer
+ * path.
+ */
+ 
+static bool GetFileFromEnv(
+  const std::string &env_name,
+  const pid_t pid,
+  const size_t path_len,
+  char *path)
+{
+  assert(path_len > 0);
+  
+  // +1 for null character at the beginning
+  // +1 for equals sign at end
+  size_t TO_FIND_LEN = env_name.length()+2;
+  
+  // Use a vector, because you can't use c++ std::string with a null character
+  // at the beginning, I tried
+  // And the vector will be cleaned up, no memory leaks!
+  std::vector<char> to_find(TO_FIND_LEN);
+  sprintf(&to_find[0], "%c%s=", '\0', env_name.c_str());
+
+  if (snprintf(path, path_len, "/proc/%d/environ", pid) >=
+      static_cast<int>(path_len))
+  {
+    if (errno == 0) {errno = ERANGE;}
+    return false;
+  }
+  int olduid = geteuid();
+  // NOTE: we ignore return values of these syscalls; this code path
+  // will work if cvmfs is FUSE-mounted as an unprivileged user.
+  seteuid(0);
+
+  FILE *fp = fopen(path, "r");
+  if (!fp) {
+    LogAuthz(kLogAuthzSyslogErr | kLogAuthzDebug,
+             "failed to open environment file for pid %d.", pid);
+    seteuid(olduid);
+    return false;
+  }
+
+  // Look for X509_USER_PROXY in the environment and store the value in path
+  int c = '\0';
+  size_t idx = 0, key_idx = 0;
+  bool set_env = false;
+  while (1) {
+    if (c == EOF) {break;}
+    if (key_idx == TO_FIND_LEN) {
+      if (idx >= path_len - 1) {break;}
+      if (c == '\0') {set_env = true; break;}
+      path[idx++] = c;
+    } else if (to_find[key_idx++] != c) {
+      key_idx = 0;
+    }
+    c = fgetc(fp);
+  }
+  fclose(fp);
+  seteuid(olduid);
+
+  if (set_env) {path[idx] = '\0';}
+  return set_env;
+}
+
+
+/**
+ * Opens a read-only file pointer to the proxy certificate as a given user.
+ * The path is either taken from X509_USER_PROXY environment from the given pid
```

Comment:
Should we have a default location?  If we do, it may be difficult since we incorporate the userid in the x509 default location.  The logic here is going to explode.

Reply:
I think the caller should provide the default location (NULL would mean “no default”).

For now, let’s not have any default SciToken location.